src/nupic/regions/SPRegion.cpp

```diff
@@ -0,0 +1,945 @@
+  args_.minPctOverlapDutyCycles = values.getScalarT<Real32>("minPctOverlapDutyCycles", 0.001f);
+  args_.dutyCyclePeriod = values.getScalarT<UInt32>("dutyCyclePeriod", 1000);
+  args_.boostStrength = values.getScalarT<Real32>("boostStrength", 0.0f);
+  args_.seed = values.getScalarT<Int32>("seed", 1);
+  args_.spVerbosity = values.getScalarT<UInt32>("spVerbosity", 0);
+  args_.wrapAround = values.getScalarT<bool>("wrapAround", true);
+  spatialImp_ = values.getString("spatialImp", "");
+
+  // variables used by this class and not passed on to the SpatialPooler class
+  args_.learningMode = (1 == values.getScalarT<UInt32>("learningMode", true));
+  args_.iter = 0;
+
+}
+
+SPRegion::SPRegion(BundleIO &bundle, Region *region) : RegionImpl(region) {
+
+  deserialize(bundle);
+}
+
+SPRegion::~SPRegion() {}
+
+void SPRegion::initialize() {
+  // Output buffers should already have been created diring initialize or deserialize.
+  Array &outputBuffer = getOutput("bottomUpOut")->getData();
+  NTA_CHECK(outputBuffer.getType() == NTA_BasicType_SDR);
+  UInt32 columnCount = (UInt32)outputBuffer.getCount();
+  if (columnCount == 0 || outputBuffer.getBuffer() == nullptr) {
+    NTA_THROW << "SPRegion::initialize - Output buffer not set.\n";
+  }
+
+  // All input links should have been initialized during Network.initialize().
+  // However, if nothing is connected it might not be. The SpatialPooler
+  // algorithm requires input.
+  //
+  // If there are more than one input link (FAN-IN), the input buffer will be the
+  // concatination of all incomming buffers.  
+  Input *in = getInput("bottomUpIn");
+  if (!in->hasIncomingLinks())
+     NTA_THROW << "SPRegion::initialize - No input links were configured for this SP region.\n";
+  Array &inputBuffer = in->getData();
+  NTA_CHECK(inputBuffer.getType() == NTA_BasicType_SDR);
+  args_.inputWidth = (UInt32)inputBuffer.getCount();
+  if (args_.inputWidth == 0) {
+    NTA_THROW << "SPRegion::initialize - No input buffer was allocated for this SP region.\n";
+  }
+
+  // Take the dimensions directly from the SDRs.
+  std::vector<UInt32> inputDimensions = inputBuffer.getSDR()->dimensions;
+  std::vector<UInt32> columnDimensions = outputBuffer.getSDR()->dimensions;
+  if (args_.potentialRadius == 0)
+    args_.potentialRadius = args_.inputWidth;
+
+  // instantiate a SpatialPooler.
+  sp_ = std::unique_ptr<SpatialPooler>(new SpatialPooler(
+      inputDimensions, columnDimensions, args_.potentialRadius,
+      args_.potentialPct, args_.globalInhibition, args_.localAreaDensity,
+      args_.numActiveColumnsPerInhArea, args_.stimulusThreshold,
+      args_.synPermInactiveDec, args_.synPermActiveInc, args_.synPermConnected,
+      args_.minPctOverlapDutyCycles, args_.dutyCyclePeriod, args_.boostStrength,
+      args_.seed, args_.spVerbosity, args_.wrapAround));
+}
+
+
+
+void SPRegion::compute() {
+  NTA_ASSERT(sp_) << "SP not initialized";
+
+  // BOTTOM-UP compute mode
+  args_.iter++;
```

Comment:
is this part of the SPRegion, doesn't SP have its own iter, which we can use?

Reply:
Hmmm, interesting....this is not used for anything :)
I think in the original .py code it was used for something that did not apply to SP.

I will remove it.